tests/ssl-opt.sh

```diff
@@ -2110,7 +2110,7 @@ run_test    "TLS: password protected server key, two certificates" \
 requires_config_enabled MBEDTLS_X509_TRUSTED_CERTIFICATE_CALLBACK
```

Comment:
Are we running the CA callback test cases both through the TLS 1.2 code path and through the TLS 1.3 code path? Looking at the outcome file, I think we are because TLS 1.3 is the default and we run these test cases in one 1.2-only configuration, namely `component_full_without_ecdhe_ecdsa_and_tls13`. Is this correct? It feels to me like the coverage is a bit fragile here. Maybe out of scope, but do we need to clarify or robustify how we test 1.2 and 1.3 code paths for features that apply to both and involve different code paths?

Reply:
On top of the robustness issue, there's also a problem with measuring coverage (lcov): we currently do that using a single build - full config, which includes 1.3. So, if there are any 1.2 code paths, `basics-build-test.sh` is going to see them as not covered.

I'm tempted to have explicit 1.2 and 1.3 versions of each test that's intended to run with both versions. That would solve both the robustness issue (if some tests are not executed with both versions, this is going to be caught by the "never executed" check that will soon be enforced) and the coverage measurement issue.

The problem is I'm not sure how to do that without duplication.

Wdyt? @gilles-peskine-arm @ronald-cron-arm

(Note: #9018 is going to use the same code path for 1.2 and 1.3 for most of the verification, but a few code paths are still going to be version-specific (extended key usage).)

Reply:
Good point about test coverage. I generally agree with testing all versions when we know that they involve different code paths.

I don't know how much code duplication it requires, but that shouldn't be too bad with proper auxiliary functions. And anyway some duplication in test code is not a major problem.

If there needs to be a lot of duplication of test cases, we have a well-established framework for generating test cases.

Reply:
> Are we running the CA callback test cases both through the TLS 1.2 code path and through the TLS 1.3 code path?

Yes, in all.sh we have at least one component where only TLS 1.2/1.3 is enabled: (test_(tls12|tls13)_only.

Reply:
> I'm tempted to have explicit 1.2 and 1.3 versions of each test that's intended to run with both versions. That would solve both the robustness issue (if some tests are not executed with both versions, this is going to be caught by the "never executed" check that will soon be enforced) and the coverage measurement issue.
> 
> The problem is I'm not sure how to do that without duplication.

It seems to be a general issue of the test coverage measurement. In all.sh we have a lot of components to test specific parts of the code and I don't think it is possible by adding more and more test cases to reach the point where with the run in one configuration we cover all the code paths that we cover with all all.sh components. To improve the test coverage measurement I would rather try to see if we can take the run of tests in several configurations. Not sure how feasible it is though.

Reply:
I agree that in general we can't cover everything with a single config (cases where some code is included when an option is set, and some other code when it's unset), but I'm not sure that's relevant here: we're talking about a case where we absolutely can cover what we want in a single config, but we are currently choosing to rely on multiple configs instead.

That choice certainly has upsides: not need to duplicate tests, and all TLS tests will (get a chance to) run with both 1.2 and 1.3 unless they explicitly force one or the other.

But it also has downsides:
- Robustness/readability as Gilles pointed out: we need to make sure we keep the 1.2-only component, but also if there are tests that have additional (negative) dependencies and need to be run with both 1.2 and 1.3, then we need a 1.2-only and a 1.3-enabled component with the additional dependency enabled/disabled (if that's not its state in full or whatever base config the 1.2-only component uses). In terms of readability, that means information is not local: if I look at a test in `ssl-opt.sh` that doesn't seem to require/force a specific version, and I want to know if it actually runs with both 1.2 and 1.3, I have to check the list of configs in `all.sh`.
- If there are different code paths for 1.2 and 1.3 (in the same build) we have no convenient way of checking that everything is covered. (We have inconvenient ways of doing so, though: measure coverage in two builds, then aggregate the results either by writing some lcov magic that does it (might be the right thing to do in the long run, but not something I've time for right now), or by just staring at both coverage results and visually merging them - tedious and error-prone.)

I think that without trying to completely address the robustness/readability issue, I'll still want to make coverage testing convenient for me when executing #9018.